pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,195 @@
+			networkToIPAMClaimName[network.Name] = IPAMClaimParams{
+				claimName:   fmt.Sprintf("%s.%s", vmiName, network.Name),
+				networkName: persistentIPsNetworkName,
+			}
+		}
+	}
+
+	return networkToIPAMClaimName, nil
+}
+
+func getPersistentIPsNetworkName(virtClient kubecli.KubevirtClient, namespace string, fullNetworkName string) (string, error) {
+	namespace, networkName := getNamespaceAndNetworkName(namespace, fullNetworkName)
+	nad, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(namespace).Get(context.Background(), networkName, v1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to locate network attachment definition %s/%s", namespace, networkName)
+	}
+
+	if nad.Spec.Config == "" {
+		return "", nil
+	}
+
+	netConf := struct {
+		AllowPersistentIPs bool   `json:"allowPersistentIPs,omitempty"`
+		Name               string `json:"name,omitempty"`
+	}{}
+	err = json.Unmarshal([]byte(nad.Spec.Config), &netConf)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal NAD spec.config JSON: %v", err)
+	}
+
+	if !netConf.AllowPersistentIPs {
+		return "", nil
+	}
+
+	if netConf.Name == "" {
+		return "", fmt.Errorf("failed to obtain network name: cannot be empty when allowPersistentIPs is true")
+	}
+
+	return netConf.Name, nil
+}
+
+func CreateIPAMClaims(client kubecli.KubevirtClient, vmi *virtv1.VirtualMachineInstance, ownerRef *v1.OwnerReference) error {
+	nonAbsentNetworks := filterNonAbsentNetworks(vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks)
+	networkToIPAMClaimParams, err := GetNetworkToIPAMClaimParams(client, vmi.Namespace, vmi.Name, nonAbsentNetworks)
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	// For a standalone VMI, the IPAMClaim owner is the VMI itself
+	if ownerRef == nil {
+		ownerRef = v1.NewControllerRef(vmi, virtv1.VirtualMachineInstanceGroupVersionKind)
+	}
+
+	claims := composeIPAMClaims(vmi.Namespace, ownerRef, networkToIPAMClaimParams)
+	err = createIPAMClaims(client, vmi.Namespace, claims)
+	if err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmi.Name, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func createIPAMClaims(client kubecli.KubevirtClient, namespace string, claims []*ipamclaims.IPAMClaim) error {
+	for _, claim := range claims {
+		_, err := client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Create(
+			context.Background(),
+			claim,
+			v1.CreateOptions{},
+		)
+
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
```

Comment:
I wonder if it wouldn't make sense to swallow this particular error ? At least if the existing claim is empty (i.e. it does **not** have IPs in its status). Like let's say it fails to create the VM for a transient reason, but somehow actually managed to create the claim for it. It would be reconciled, and would fail here until the end of times .

You're already reading the Claim below (in the validate func). This would be changing the operation order.

Reply:
Is it for errors different than `IsAlreadyExists`.
Only if Create fails on "Already Exists" we try to Get and validate
If there error is something else (i.e etcd timeout), we don't need to try and Get, but wait for the next reconcile

In the scenario you mentioned if it is already created, then here it will eventually fail on "IsAlreadyExists"
and proceed to the validation, so it won't fail until end of times.